### PR TITLE
Filecaseexist

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

File: src/d_iwad.c [hunk 1]
```diff
@@ -459,13 +459,14 @@ static boolean DirIsFile(char *path, char *filename)
 static char *CheckDirectoryHasIWAD(char *dir, char *iwadname)
 {
     char *filename; 
+    char *probe;
 
     // As a special case, the "directory" may refer directly to an
     // IWAD file if the path comes from DOOMWADDIR or DOOMWADPATH.
 
-    if (DirIsFile(dir, iwadname) && M_FileExists(dir))
+    if (DirIsFile(dir, iwadname) && (probe = M_FileCaseExists(dir)))
```
Comment (fragglet) [inline src/d_iwad.c]: Can you please refactor to not do assignments inside `if()` statements?

File: src/d_iwad.c [hunk 2]
```diff
@@ -459,13 +459,15 @@ static boolean DirIsFile(char *path, char *filename)
...
 
     // As a special case, the "directory" may refer directly to an
     // IWAD file if the path comes from DOOMWADDIR or DOOMWADPATH.
 
-    if (DirIsFile(dir, iwadname) && M_FileExists(dir))
+    probe = M_FileCaseExists(dir);
+    if (DirIsFile(dir, iwadname) && (probe != NULL))
     {
-        return M_StringDuplicate(dir);
+        return M_StringDuplicate(probe);
```
Comment (fragglet) [inline src/d_iwad.c]: `M_FileCaseExists()` allocates a new string, so no need for `M_StringDuplicate()` here too, correct? ie. this is a memory leak?

File: src/d_iwad.c [hunk 3]
```diff
@@ -459,13 +459,15 @@ static boolean DirIsFile(char *path, char *filename)
...
 {
     char *filename; 
+    char *probe;
 
     // As a special case, the "directory" may refer directly to an
     // IWAD file if the path comes from DOOMWADDIR or DOOMWADPATH.
 
-    if (DirIsFile(dir, iwadname) && M_FileExists(dir))
+    probe = M_FileCaseExists(dir);
+    if (DirIsFile(dir, iwadname) && (probe != NULL))
```
Comment (fragglet) [inline src/d_iwad.c]: No need for extra parens.

File: src/m_misc.c [hunk 4]
```diff
@@ -83,6 +83,72 @@ boolean M_FileExists(char *filename)
     }
 }
 
+char *M_FileCaseExists(char *path)
```
Comment (fragglet) [inline src/m_misc.c]: Can you add a short comment before this function documenting its behavior and specifically the fact that it returns a newly allocated string that the caller is responsible for freeing?

File: src/m_misc.c [hunk 5]
```diff
@@ -250,6 +316,16 @@ void M_ForceUppercase(char *text)
     }
 }
 
+void M_ForceLowercase(char *text)
```
Comment (fragglet) [inline src/m_misc.c]: This could do with a doc string too.

File: src/m_misc.c (hunk 4 above)
Comment (fragglet) [inline src/m_misc.c]: Oh. Then please change it to always return a newly-allocated string. Simpler function semantics are better/safer.

Comment (fabiangreffrath) [inline src/m_misc.c]: Should be consistent now...

**Comment by fragglet** [comment]
Thanks!

**Pull request merged**